File: algo-anagrams-i/js/characterMatch.js
#!/usr/bin/env node

/**
 * 
 * @param {string} string1 Target string to check for a match
 * @param {string} string2 String used to check from
 * @returns Bool - True if a match is found, False if not
 */

function isCharacterMatch(string1, string2) {

    // Construct a RegEx with the characters from strCheck
    // Flagged as case insensitive
    let regexCheckPattern = new RegExp(`[${string2}]{${string2.length}}`,"i");

    //console.log(regexCheckPattern)

    return regexCheckPattern.test(string1)
}

function anagramsFor(word, listOfWords) {
    let arrResult = []

    for(let strThisWord of listOfWords) {
        if(isCharacterMatch(word, strThisWord)) {
            arrResult.push(strThisWord)
        }

    }

    return arrResult;
}

exports.isCharacterMatch = function(string1, string2) {
    isCharacterMatch(string1, string2)
};

exports.anagramsFor = function(word, listOfWords) {
    anagramsFor(word, listOfWords)
};

//module.exports = isCharacterMatch, anagramsFor;

let listOfWords = ["threads", "trashed", "hardest", "hatreds", "hounds"];

//console.log(isCharacterMatch("CharM", "mARch"))
console.log(anagramsFor("threads", listOfWords))
console.log(anagramsFor("threads", listOfWords).length)

File: algo/algo-anagrams-i/python/character_match.py
#!/usr/bin/env python3.12
# Don't forget to run the tests (and create some of your own)
import re

# Part 1
def is_character_match(string1, string2):
	# Construct the string to feed the RegEx pattern
	# Given the string1 as 'charm' this will create the RegEx
	# pattern r'[charm]{5}' with the case insenstivie flag applied
	# This will ensure we find exactly 5 matches of the characters listed in the
	# string
	str_pattern = f"[{string1}]" + "{" + f"{len(string1)}" + "}"

	regex_pattern = re.compile(str_pattern, re.I)

	if re.match(regex_pattern,string2) :
		return True
	return False


# Part 2
def anagrams_for(word, list_of_words):
	lst_result = []
	while list_of_words :
		# Use the match function we've already created
		if is_character_match(word, list_of_words[0]) :
			# if we have a match, add it to the results list
			lst_result.append(list_of_words[0])
		
		# remove the element from the original list
		del(list_of_words[0])
		
	return lst_result

# list_of_words = ["threads", "trashed", "hardest", "hatreds", "hounds"]

# print(is_character_match('charm', 'march'))
# print(is_character_match('zach','attack'))

# print(anagrams_for("threads", list_of_words))
# print(anagrams_for("apple", list_of_words))